test/test-tcp-write-ready.c

```diff
@@ -0,0 +1,210 @@
+#endif
+
+#ifndef _WIN32
+  {
+    /* Allow reuse of the port. */
+    int yes = 1;
+    int r = setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof yes);
+    ASSERT(r == 0);
+  }
+#endif
+
+  return sock;
+}
+
+
+static void close_socket(uv_os_sock_t sock) {
+  int r;
+#ifdef _WIN32
+  r = closesocket(sock);
+#else
+  r = close(sock);
+#endif
+  ASSERT(r == 0);
+}
+
+
+static void alloc_cb(uv_handle_t* handle,
+                     size_t suggested_size,
+                     uv_buf_t* buf) {
+  static char slab[65536];
+  ASSERT(suggested_size <= sizeof(slab));
+  buf->base = slab;
+  buf->len = sizeof(slab);
+}
+
+
+static void close_cb(uv_handle_t* handle) {
+  ASSERT(handle != NULL);
+  close_cb_called++;
+}
+
+
+static void shutdown_cb(uv_shutdown_t* req, int status) {
+  ASSERT(req == &shutdown_req);
+  ASSERT(status == 0);
+
+  /* Now we wait for the EOF */
+  shutdown_cb_called++;
+}
+
+
+static void read_cb(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf) {
+  ASSERT(tcp != NULL);
+
+  if (nread >= 0) {
+    for (int i = 0; i < nread; ++i)
+      ASSERT(buf->base[i] == 'P');
+  }
+  else {
+    ASSERT(nread == UV_EOF);
+    printf("GOT EOF\n");
+    uv_close((uv_handle_t*)tcp, close_cb);
+  }
+}
+
+
+static void write_cb(uv_write_t* req, int status) {
+  uv_buf_t buf = uv_buf_init("P", 1);
+  ASSERT(req != NULL);
+  int r;
+
+  if (status) {
+    ASSERT(shutdown_cb_called);
+    return;
+  }
+
+  r = uv_write(&write_req, req->handle, &buf, 1, write_cb);
+  ASSERT(r == 0);
+
+  write_cb_called++;
+}
+
+
+static void timer_cb(uv_timer_t* handle) {
+  int r;
+
+  /* Shutdown on drain. */
+  r = uv_shutdown(&shutdown_req, &client, shutdown_cb);
+  ASSERT(r == 0);
+}
+
+
+static void connect_cb(uv_connect_t* req, int status) {
+  uv_buf_t buf = uv_buf_init("P", 1);
+  uv_stream_t* stream;
+  int r;
+
+  ASSERT(req == &connect_req);
+  ASSERT(status == 0);
+
+  stream = req->handle;
+  connect_cb_called++;
+
+  r = uv_timer_init(uv_default_loop(), &tm);
+  ASSERT(r == 0);
+
+  r = uv_timer_start(&tm, timer_cb, 2000, 0);
+  ASSERT(r == 0);
+
+  r = uv_write(&write_req, stream, &buf, 1, write_cb);
+  ASSERT(r == 0);
+
+  /* Start reading */
+  r = uv_read_start(stream, alloc_cb, read_cb);
+  ASSERT(r == 0);
+}
+
+
+TEST_IMPL(tcp_write_ready) {
+  struct sockaddr_in addr;
+  uv_os_sock_t sock;
+  int r;
+
+  ASSERT(0 == uv_ip4_addr("127.0.0.1", TEST_PORT, &addr));
+
+  startup();
+  sock = create_tcp_socket();
+
+  r = uv_tcp_init(uv_default_loop(), &client);
+  ASSERT(r == 0);
+
+  r = uv_tcp_open(&client, sock);
+  ASSERT(r == 0);
+
+  r = uv_tcp_connect(&connect_req,
+                     &client,
+                     (const struct sockaddr*) &addr,
+                     connect_cb);
+  ASSERT(r == 0);
+
+  uv_run(uv_default_loop(), UV_RUN_DEFAULT);
+
+  ASSERT(shutdown_cb_called == 1);
+  ASSERT(connect_cb_called == 1);
+  ASSERT(write_cb_called > 0);
+  ASSERT(close_cb_called == 1);
+
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
```

Comment:
Can you merge this into `test/test-tcp-open.c`?  Duplicates too much code otherwise, IMO.

Reply:
good idea. Will do

Reply:
Done